### PR TITLE
[Frontend] Add participant team approval feature by host

### PR DESCRIPTION
**Pull request by Suryansh5545**
This PR will allow the host to decide wheather he want to manually approve participant team and if yes then the host will be shown a Team Approved View as shown in the video below

https://github.com/Cloud-CV/EvalAI/assets/34577232/e2c35d02-ccad-48ef-b61f-bf307fd2f11d

### CODE REVIEW

**Review by gautamjajoo — CHANGES_REQUESTED** [review]
A couple of things:
1. The checkbox is placed weirdly. Can you add a similar margin on the card's left side and center the box?
2. a table scroll appears; attaching the screenshot as well.
3. The team created by the host should be automatically approved. @gchhablani, can you also share your opinion?
4. Can the tab `Team Approved` and the title `Team Approval` be the same? @gchhablani, what should be the name of the tab?
5. Can we approve the team in a similar way we verify submission on all submissions page?

File: apps/challenges/challenge_config_utils.py [hunk 1]
```diff
@@ -311,6 +311,13 @@ def validate_challenge_config_util(
             yaml_file_data, challenge_config_location, "description"
         )
 
+    # Check Manual Approval
+    if "maunal_approval" in yaml_file_data:
```
Comment (gchhablani) [inline apps/challenges/challenge_config_utils.py:315]: Make this `manual_participant_approval` everywhere.

File: apps/challenges/migrations/0091_auto_manual_approval.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
+# Generated by Django 2.2.20 on 2023-06-09 12:57
+
+from django.db import migrations, models
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('participants', '0012_remove_docker_repository_uri_from_team'),
+        ('challenges', '0090_challenge_allow_resuming_submissions'),
```
Comment (gchhablani) [inline apps/challenges/migrations/0091_auto_manual_approval.py:10]: You will need to update the migration here. Delete this and generate migration again.

File: apps/challenges/views.py [hunk 3]
```diff
@@ -506,6 +523,75 @@ def add_participant_team_to_challenge(
...
+            submission_exist = Submission.objects.filter(participant_team=participant_team, challenge_phase=challenge_phase).exists()
+            if submission_exist:
+                break
+    else:
+        submission_exist = False
+    if challenge.approved_participant_teams.filter(pk=participant_team_pk).exists() and not submission_exist:
+        challenge.approved_participant_teams.remove(participant_team)
+        return Response(status=status.HTTP_204_NO_CONTENT)
+    else:
+        response_data = {"error": "Participant Team not approved or already have a Submission"}
```
Comment (gchhablani) [inline apps/challenges/views.py:591]: "Participant Team {name} was not approved"
"Participant Team {name} has existing submissions and cannot be unapproved"

^ Make these cases separate and use this text.

File: apps/challenges/views.py [hunk 4]
```diff
@@ -385,7 +395,14 @@ def get_participant_teams_for_challenge(request, challenge_pk):
         serializer = ParticipantTeamDetailSerializer(
             participant_teams, many=True
         )
-        response_data = serializer.data
+        for participant_team in serializer.data:
+            if (challenge.approved_participant_teams.filter(id=participant_team["id"]).exists()):
+                participant_team["approved"] = True
```
[the remark this answers is not in the document]
Comment (Suryansh5545) [reply apps/challenges/views.py:400]: in getTeamName function in ChallengeCtrl. to verify if manual approval it on and if the team is approved or not. if not approved the submission will not be allowed to submitted

File: apps/challenges/views.py (hunk 3 above)
Comment (Suryansh5545) [reply apps/challenges/views.py:591]: Done

File: frontend/src/js/controllers/challengeCtrl.js [hunk 5]
```diff
@@ -78,6 +78,7 @@
         vm.currentPhaseMetaAttributesVisibility = {};
         vm.phaseLeaderboardPublic = [];
         vm.currentPhaseLeaderboardPublic = false;
+        vm.approved_by_host = false;
```
Comment (gchhablani) [inline frontend/src/js/controllers/challengeCtrl.js:81]: The variable `approved_by_host` does not make sense. Not all challenges need approval from host. Make this variable `eligible_to_submit` or something.

File: frontend/src/js/controllers/challengeCtrl.js [hunk 6]
```diff
@@ -2835,5 +2920,6 @@
         };
 
     }
+
```
Comment (gchhablani) [inline frontend/src/js/controllers/challengeCtrl.js:2923]: Remove this line.

File: frontend/src/views/web/challenge/edit-challenge/edit-challenge-approvalConfirm.html [hunk 7]
```diff
@@ -0,0 +1,20 @@
+<section class="text-center pd-20 width-900">
+    <div class="row ht-wt-100p">
+        <div class="col s12 m12">
+            <div class="ev-card-body">
+                <form name="approvalform" ng-submit="challenge.check_approval_status(challenge.dialog.challengeId, challenge.dialog.participant_team_id, challenge.dialog.approved_status, approvalform.$valid)">
+                    <div class="pass-title">Approve Team</div>
+                    <div class="pass-title"><b>Participant Team with submission will no longer be disapproved</b></div>
```
Comment (gchhablani) [inline frontend/src/views/web/challenge/edit-challenge/edit-challenge-approvalConfirm.html:7]: <b>Note</b>: Participant teams with existing submissions cannot be disapproved

File: frontend/tests/controllers-test/challengeCtrl.test.js [hunk 8]
```diff
@@ -633,6 +633,7 @@ describe('Unit tests for challenge controller', function () {
                 ]
             };
             success = true;
+            vm.approved_by_host = true;
```
Comment (gchhablani) [inline frontend/tests/controllers-test/challengeCtrl.test.js:636]: `eligible_to_submit`

File: frontend/src/views/web/challenge/submission.html [hunk 9]
```diff
@@ -28,6 +28,8 @@ <h5 class="w-300">Submission Guidelines</h5>
     </div>
 </section>
 
+<!-- To Check if Manual Approval is done or off-->
+<div ng-if="challenge.approved_by_host">
```
Comment (gchhablani) [inline frontend/src/views/web/challenge/submission.html:32]: `eligible_to_submit`

File: tests/unit/challenges/test_views.py [hunk 10]
```diff
@@ -231,6 +233,82 @@ def test_team_name_for_challenge_with_participant_team_does_not_exist(
         self.assertEqual(response.status_code, status.HTTP_406_NOT_ACCEPTABLE)
 
 
+class GetApprovedParticipantTeamNameTest(BaseAPITestClass):
```
Comment (gchhablani) [inline tests/unit/challenges/test_views.py:236]: Excellent! Makes this a complete PR!

**Review by gchhablani — COMMENTED** [review]
This PR is ready to go!
Just added a few comments and questions.
Once those are resolved, we can merge and test it.

File: apps/challenges/views.py [hunk 11]
```diff
@@ -331,12 +331,22 @@ def participant_team_detail_for_challenge(request, challenge_pk):
...
-        response_data = serializer.data
+        if (challenge.approved_participant_teams.filter(pk=participant_team_pk).exists()):
+            approved = True
+        elif not challenge.manual_participant_approval:
+            approved = True
+        else:
+            approved = False
+        response_data = {
+            "approved": approved,
+            "participant_team": serializer.data,
```
[the remark this answers is not in the document]
Comment (Suryansh5545) [reply apps/challenges/views.py:348]: it is used by getTeamName in ChallengeCtl line 272 only

File: apps/challenges/views.py [hunk 12]
```diff
@@ -385,7 +395,14 @@ def get_participant_teams_for_challenge(request, challenge_pk):
         serializer = ParticipantTeamDetailSerializer(
             participant_teams, many=True
         )
-        response_data = serializer.data
+        for participant_team in serializer.data:
+            if (challenge.approved_participant_teams.filter(id=participant_team["id"]).exists()):
```
[the remark this answers is not in the document]
Comment (Suryansh5545) [reply apps/challenges/views.py:399]: This endpoint is only used when challenge is manually approved, used to get the list of participant team and there approval status

File: apps/challenges/migrations/0093_auto_manual_approval.py [hunk 13]
```diff
@@ -0,0 +1,24 @@
+# Generated by Django 2.2.20 on 2023-06-09 12:57
+
+from django.db import migrations, models
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('participants', '0012_remove_docker_repository_uri_from_team'),
+        ('challenges', '0092_challenge_allow_cancel_running_submissions'),
```
[the remark this answers is not in the document]
Comment (Suryansh5545) [reply apps/challenges/migrations/0093_auto_manual_approval.py]: Yes

**Pull request merged**